### PR TITLE
Use `fetch` rather than `XMLHttpRequest` for downloading Wasm files

### PR DESCRIPTION
**Pull request by seanmorris**
This PR switches requests for Wasm files to be loaded with `fetch()` rather than `XMLHttpRequest()`.

This fixes running in service workers where `XMLHttpRequest` is not available.

Partially fixes https://github.com/emscripten-core/emscripten/issues/22003, only applies to Wasm files.

### CODE REVIEW

**Comment by sbc100** [comment]
I believe `readAsync` is also used to load the wasm module itself.. so this effects all users, not just those who have shared libraries.  Maybe update the title and description to refer to wasm file loading?

File: src/web_or_worker_shell_read.js [hunk 1]
```diff
@@ -22,17 +22,12 @@
...
-    var xhr = new XMLHttpRequest();
-    xhr.open('GET', url, true);
-    xhr.responseType = 'arraybuffer';
-    xhr.onload = () => {
-      if (xhr.status == 200 || (xhr.status == 0 && xhr.response)) { // file URLs can return 0
-        onload(xhr.response);
-        return;
+    fetch(url)
+    .then(response => {
+      if(response.ok) {
```
Comment (sbc100) [inline src/web_or_worker_shell_read.js:27]: Space after `if`

File: src/web_or_worker_shell_read.js [hunk 2]
```diff
@@ -22,17 +22,12 @@
...
   readAsync = (url, onload, onerror) => {
-    var xhr = new XMLHttpRequest();
-    xhr.open('GET', url, true);
-    xhr.responseType = 'arraybuffer';
-    xhr.onload = () => {
-      if (xhr.status == 200 || (xhr.status == 0 && xhr.response)) { // file URLs can return 0
-        onload(xhr.response);
-        return;
+    fetch(url)
+    .then(response => {
```
Comment (sbc100) [inline src/web_or_worker_shell_read.js:26]: Add brackets around single argument to arrow function

File: src/web_or_worker_shell_read.js [hunk 3]
```diff
@@ -22,17 +22,12 @@
...
 
   readAsync = (url, onload, onerror) => {
-    var xhr = new XMLHttpRequest();
-    xhr.open('GET', url, true);
-    xhr.responseType = 'arraybuffer';
-    xhr.onload = () => {
-      if (xhr.status == 200 || (xhr.status == 0 && xhr.response)) { // file URLs can return 0
-        onload(xhr.response);
-        return;
+    fetch(url)
```
Comment (sbc100) [inline src/web_or_worker_shell_read.js:25]: Can you also update the above two uses of `XMLHttpRequest` ?

File: test/test_browser.py [hunk 4]
```diff
@@ -5549,6 +5554,69 @@ def test_webpack(self):
...
+                  xhr.open('GET', 'http://localhost:8888/report_result?' + result, true);
+                  xhr.send();
+                };
+              })(),
+              canvas: document.getElementById('canvas')
+            };
+          </script>
+          {{{ SCRIPT }}}
+        </body>
+      </html>''')
```
Comment (sbc100) [inline test/test_browser.py:5603]: Is this file needed for something?  I would hope we can simply avoid this and use the default shell?

File: test/test_browser.py [hunk 5]
```diff
@@ -5549,6 +5554,69 @@ def test_webpack(self):
...
+      int main() {
+        int found = EM_ASM_INT(
+          return preloadedWasm['/library.so'] !== undefined;
+        );
+        void *lib_handle = dlopen("/library.so", RTLD_NOW);
+        typedef int (*voidfunc)();
+        voidfunc x = (voidfunc)dlsym(lib_handle, "library_func");
+        printf("Got val: %d\n", x());
+        assert(x() == 42);
+        return 0;
```
Comment (sbc100) [inline test/test_browser.py:5578]: Just return 42 here and then it will show up as `/report_result?42` below I think.  No need for a custom shell html.

**Review by sbc100 — COMMENTED** [review]
lgtm with a few comments

File: test/test_browser.py [hunk 6]
```diff
@@ -5549,6 +5554,69 @@ def test_webpack(self):
     shutil.copyfile('webpack/src/hello.wasm', 'webpack/dist/hello.wasm')
     self.run_browser('webpack/dist/index.html', '/report_result?exit:0')
 
+  def test_fetch_polyfill_shared_lib(self):
```
Comment (sbc100) [inline test/test_browser.py:5557]: I wonder if we even need to involve shared libraries here since the polyfill is needed even to load the base wasm file (I think).

**Review by sbc100 — COMMENTED** [review]
I updated the PR title and description.. I hope thats ok

**Comment by sbc100** [comment]
Actually I was wrong.. it looks like the core wasm file already has a fetch-by-default setup (we can remove this as a followup to this change I hope): https://github.com/emscripten-core/emscripten/blob/cb99414efed02dc61d04315d3e3cf5ad3180e56f/src/preamble.js#L660

File: src/web_or_worker_shell_read.js (hunk 3 above)
Comment (seanmorris) [reply src/web_or_worker_shell_read.js:25]: Looks like the two earlier calls are synchronous, it might take some more re-engineering to get those converted.

File: test/test_browser.py (hunk 4 above)
Comment (seanmorris) [reply test/test_browser.py:5603]: I need that `print` function on like 5587-5606 to report the result from WASM to the testing infrastructure.

File: test/test_browser.py (hunk 5 above)
Comment (seanmorris) [reply test/test_browser.py:5578]: Interesting! I'll give it a shot.

Comment (seanmorris) [reply test/test_browser.py:5577]: I wasn't able to get a result in the test without using the custom shell. I reduced the HTML to the bare-minimum.

File: test/test_browser.py (hunk 6 above)
Comment (seanmorris) [reply test/test_browser.py:5557]: I think its good to have this test either way, that stuff could be changed in the future as the code becomes more unified. We should make sure the code works in all cases.

File: test/common.py [hunk 7]
```diff
@@ -2270,7 +2270,7 @@ def compile_btest(self, filename, args, reporting=Reporting.FULL):
       if reporting == Reporting.FULL:
         # If C reporting (i.e. the REPORT_RESULT macro) is required we
         # also include report_result.c and force-include report_result.h
-        self.run_process([EMCC, '-c', '-I' + TEST_ROOT,
+        self.run_process([EMCC, '-c', '-fPIC', '-I' + TEST_ROOT,
```
Comment (sbc100) [inline test/common.py:2273]: Hmm.. I wonder why this is needed now and not for any other existing dynamic linking browser tests.    I will take a look.

File: test/test_browser.py [hunk 8]
```diff
@@ -699,15 +699,20 @@ def setup(assetLocalization):
...
               window.onerror = null;
               var result = error.includes("test.data") ? 1 : 0;
               var xhr = new XMLHttpRequest();
               xhr.open('GET', 'http://localhost:8888/report_result?' + result, true);
               xhr.send();
+              console.log(event, error);
               setTimeout(function() { window.close() }, 1000);
             }
+            window.addEventListener('error', handler);
+            window.addEventListener('unhandledrejection', handler);
```
Comment (sbc100) [inline test/test_browser.py:715]: Is this change needed because we now generate `unhandledrejection` in more cases than before?

File: test/test_browser.py [hunk 9]
```diff
@@ -5549,6 +5554,58 @@ def test_webpack(self):
...
+      #include <stdio.h>
+      #include <emscripten.h>
+      int main() {
+        int found = EM_ASM_INT(
+          return preloadedWasm['/library.so'] !== undefined;
+        );
+        void *lib_handle = dlopen("/library.so", RTLD_NOW);
+        typedef int (*voidfunc)();
+        voidfunc x = (voidfunc)dlsym(lib_handle, "library_func");
+        printf("%d\n", x());
```
Comment (sbc100) [inline test/test_browser.py:5576]: Another way to avoid the explicit shell is simply to use the `REPORT_RESULT` macro here in the C code.

File: test/test_browser.py [hunk 10]
```diff
@@ -5549,6 +5554,58 @@ def test_webpack(self):
...
+    self.run_process([EMCC, 'library.c', '-sSIDE_MODULE', '-O2', '-o', 'library.so'])
+
+    def test(args, expect_fail):
+      self.compile_btest('main.c', ['library.so', '-sMAIN_MODULE', '--shell-file', 'report_result.html', '-o', 'a.out.html'])
+      js = read_file('a.out.js')
+      if expect_fail:
+        create_file('a.out.js', 'fetch = undefined;\n' + js)
+        return self.run_browser('a.out.html', '/report_result?abort:TypeError')
+      else:
+         return self.run_browser('a.out.html', '/report_result?1')
```
Comment (sbc100) [inline test/test_browser.py:5602]: indentation looks wrong here

File: test/test_browser.py [hunk 11]
```diff
@@ -5549,6 +5554,58 @@ def test_webpack(self):
...
+        </script>
+        {{{ SCRIPT }}}
+      </body>
+      </html>
+    ''')
+
+    self.run_process([EMCC, 'library.c', '-sSIDE_MODULE', '-O2', '-o', 'library.so'])
+
+    def test(args, expect_fail):
+      self.compile_btest('main.c', ['library.so', '-sMAIN_MODULE', '--shell-file', 'report_result.html', '-o', 'a.out.html'])
```
Comment (sbc100) [inline test/test_browser.py:5596]: Can you use `-sMAIN_MODULE=2` here?   This makes the test program a lot smaller.

File: test/test_browser.py (hunk 5 above)
Comment (seanmorris) [reply test/test_browser.py:5577]: If I remove the `'--shell-file', 'report_result.html'` and return from `main()` then the test just hangs:

```
Test suites:
['test_browser']
Running test_browser: (1 tests)
(checking sanity from test runner)
shared:INFO: (Emscripten: Running sanity checks)
[Browser harness server on process 3558299]
common:INFO: No EMTEST_BROWSER set. Defaulting to `google-chrome`
common:INFO: Launching browser: ['google-chrome']

Running the browser tests. Make sure the browser allows popups from localhost.

test_fetch_polyfill_shared_lib (test_browser.browser.test_fetch_polyfill_shared_lib) ... Opening in existing browser session.
[unresponsive tests: 1]
common:INFO: Restarting browser process
common:INFO: Launching browser: ['google-chrome']
[test error (see below), automatically retrying]
Expected to find '/report_result?1
' in '[no http server activity]
', diff:

--- expected
+++ actual
@@ -1 +1 @@
-/report_result?1
+[no http server activity]
```

File: test/test_browser.py [hunk 12]
```diff
@@ -5549,6 +5554,58 @@ def test_webpack(self):
...
+    ''')
+    create_file('main.c', r'''
+      #include <assert.h>
+      #include <dlfcn.h>
+      #include <stdio.h>
+      #include <emscripten.h>
+      int main() {
+        int found = EM_ASM_INT(
+          return preloadedWasm['/library.so'] !== undefined;
+        );
```
Comment (sbc100) [inline test/test_browser.py:5572]: I don't think that above three lines should be needed as part of this test.  Unless I'm misunderstanding something?

File: test/test_browser.py (hunk 5 above)
Comment (sbc100) [reply test/test_browser.py:5577]: Can you try adding `-sEXIT_RUNTIME`?

File: test/test_browser.py (hunk 8 above)
Comment (seanmorris) [reply test/test_browser.py:715]: Throwing an error inside a promise, or rejecting for any other reason will generate ONLY an `unhandledrejection` event, not an `error` event. Unless of course you `catch()` it.

File: test/test_browser.py (hunk 5 above)
Comment (seanmorris) [reply test/test_browser.py]: > Can you try adding `-sEXIT_RUNTIME`?

That worked!!!

File: test/test_browser.py [hunk 13]
```diff
@@ -5549,6 +5554,41 @@ def test_webpack(self):
...
+      #include <stdio.h>
+      int library_func() {
+        return 42;
+      }
+    ''')
+    create_file('main.c', r'''
+      #include <assert.h>
+      #include <dlfcn.h>
+      #include <stdio.h>
+      #include <emscripten.h>
```
Comment (sbc100) [inline test/test_browser.py:5568]: Remove this header.

File: test/test_browser.py [hunk 14]
```diff
@@ -5549,6 +5554,41 @@ def test_webpack(self):
...
+        voidfunc x = (voidfunc)dlsym(lib_handle, "library_func");
+        return x();
+      }
+    ''')
+
+    self.run_process([EMCC, 'library.c', '-sSIDE_MODULE', '-O2', '-o', 'library.so'])
+
+    def test(args, expect_fail):
+      self.compile_btest('main.c', ['library.so', '-sMAIN_MODULE=2', '-sEXIT_RUNTIME', '-o', 'a.out.html'])
+      js = read_file('a.out.js')
```
Comment (sbc100) [inline test/test_browser.py:5581]: You can move this line inside the `expect_fail` block

**Review by sbc100 — COMMENTED** [review]
Looks good!

Just need to figure out what is going on with that common.py change.  Otherwise this looks ready to land. 

Thanks for you patience on this one.

File: test/test_browser.py [hunk 15]
```diff
@@ -699,7 +699,9 @@ def setup(assetLocalization):
           <center><canvas id='canvas' width='256' height='256'></canvas></center>
           <hr><div id='output'></div><hr>
           <script type='text/javascript'>
-            window.onerror = (error) => {
+            const handler = event => {
```
Comment (sbc100) [inline test/test_browser.py:702]: Braces around `event` argument here

Comment (seanmorris) [reply test/test_browser.py]: Done.

File: test/common.py (hunk 7 above)
Comment (sbc100) [reply test/common.py:2273]: Instead of doing this can you do `+ args` at the end of this command line?

Comment (sbc100) [reply test/common.py:2273]: (you may also need to add `-Wno-unused`)

Comment (seanmorris) [reply test/common.py:2273]: No problem.

Comment (seanmorris) [reply test/common.py:2273]: @sbc100 Looks like `args` doesn't get passed into this call:
https://github.com/emscripten-core/emscripten/blob/16559db15997f6de648c9107e26033043f8a5313/test/common.py#L2273-L2275

Comment (sbc100) [reply test/common.py:2273]: Yes, I suggest just adding `+ args` after the `+ self.get_emcc_args(compile_only=True)`

Comment (seanmorris) [reply test/common.py]: Some of the args passed to `compile_btest()` will interfere with that. I've added detection for `-fPIC` and am adding ONLY that flag if its present in `args`:

https://github.com/emscripten-core/emscripten/blob/feab2d60ef886ed6b07ae6cf187d312c2d2b65ad/test/common.py#L2275

Comment (sbc100) [reply test/common.py]: Which argument interfere?  Is it your test or some other test where the interference occurs?

Comment (seanmorris) [reply test/common.py]: My test fails if I just do `+ args`:

```
Test suites:
['test_browser']
Running test_browser: (1 tests)
(checking sanity from test runner)
shared:INFO: (Emscripten: Running sanity checks)
[Browser harness server on process 152289]
common:INFO: No EMTEST_BROWSER set. Defaulting to `google-chrome`
common:INFO: Launching browser: ['google-chrome']

Running the browser tests. Make sure the browser allows popups from localhost.

test_fetch_polyfill_shared_lib (test_browser.browser.test_fetch_polyfill_shared_lib) ... Opening in existing browser session.
emcc: error: linker setting ignored during compilation: 'EXIT_RUNTIME' [-Wunused-command-line-argument] [-Werror]
None
None
FAIL
[Browser harness server terminated]

======================================================================
FAIL: test_fetch_polyfill_shared_lib (test_browser.browser.test_fetch_polyfill_shared_lib)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sean/projects/php-wasm/emscripten/test/common.py", line 1550, in run_process
    return shared.run_process(cmd, check=check, **args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sean/projects/php-wasm/emscripten/tools/shared.py", line 130, in run_process
    ret = subprocess.run(cmd, check=check, input=input, *args, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/sean/projects/php-wasm/emscripten/emcc', '-c', '-I/home/sean/projects/php-wasm/emscripten/test', '-DEMTEST_PORT_NUMBER=8888', '/home/sean/projects/php-wasm/emscripten/test/report_result.c', '-Wclosure', '-Werror', '-Wno-limited-postlink-optimizations', '-Wno-pointer-sign', '-Wno-int-conversion', '-fPIC', 'library.so', '-sMAIN_MODULE=2', '-sEXIT_RUNTIME', '-o', 'a.out.html', '--pre-js', '/home/sean/projects/php-wasm/emscripten/test/browser_reporting.js']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sean/projects/php-wasm/emscripten/test/test_browser.py", line 5585, in test_fetch_polyfill_shared_lib
    test([], expect_fail=True)
  File "/home/sean/projects/php-wasm/emscripten/test/test_browser.py", line 5577, in test
    self.compile_btest('main.c', ['-fPIC', 'library.so', '-sMAIN_MODULE=2', '-sEXIT_RUNTIME', '-o', 'a.out.html'] + args)
  File "/home/sean/projects/php-wasm/emscripten/test/common.py", line 2273, in compile_btest
    self.run_process([EMCC, '-c', '-I' + TEST_ROOT,
  File "/home/sean/projects/php-wasm/emscripten/test/common.py", line 1555, in run_process
    self.fail(f'subprocess exited with non-zero return code({e.returncode}): `{shared.shlex_join(cmd)}`')
AssertionError: subprocess exited with non-zero return code(1): `/home/sean/projects/php-wasm/emscripten/emcc -c -I/home/sean/projects/php-wasm/emscripten/test -DEMTEST_PORT_NUMBER=8888 /home/sean/projects/php-wasm/emscripten/test/report_result.c -Wclosure -Werror -Wno-limited-postlink-optimizations -Wno-pointer-sign -Wno-int-conversion -fPIC library.so -sMAIN_MODULE=2 -sEXIT_RUNTIME -o a.out.html --pre-js /home/sean/projects/php-wasm/emscripten/test/browser_reporting.js`

----------------------------------------------------------------------
Ran 1 test in 0.364s

FAILED (failures=1)
```

Comment (sbc100) [reply test/common.py]: You can add `-Wno-unused-command-line-argument` too.. since its reasonable to ignore linker-only flags here.

Comment (seanmorris) [reply test/common.py]: With: `'-Wno-unused-command-line-argument`:

```
Test suites:
['test_browser']
Running test_browser: (1 tests)
(checking sanity from test runner)
shared:INFO: (Emscripten: Running sanity checks)
[Browser harness server on process 156229]
common:INFO: No EMTEST_BROWSER set. Defaulting to `google-chrome`
common:INFO: Launching browser: ['google-chrome']

Running the browser tests. Make sure the browser allows popups from localhost.

test_fetch_polyfill_shared_lib (test_browser.browser.test_fetch_polyfill_shared_lib) ... Opening in existing browser session.
emcc: error: report_result.o: No such file or directory ("report_result.o" was expected to be an input file, based on the commandline arguments provided)
None
None
FAIL
[Browser harness server terminated]

======================================================================
FAIL: test_fetch_polyfill_shared_lib (test_browser.browser.test_fetch_polyfill_shared_lib)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sean/projects/php-wasm/emscripten/test/common.py", line 1550, in run_process
    return shared.run_process(cmd, check=check, **args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sean/projects/php-wasm/emscripten/tools/shared.py", line 130, in run_process
    ret = subprocess.run(cmd, check=check, input=input, *args, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/sean/projects/php-wasm/emscripten/emcc', 'main.c', '-sNO_DEFAULT_TO_CXX', '-Wclosure', '-Werror', '-Wno-limited-postlink-optimizations', '-Wno-pointer-sign', '-Wno-int-conversion', '-fPIC', '-Wno-unused-command-line-argument', 'library.so', '-sMAIN_MODULE=2', '-sEXIT_RUNTIME', '-o', 'a.out.html', '--pre-js', '/home/sean/projects/php-wasm/emscripten/test/browser_reporting.js', 'report_result.o', '-include', '/home/sean/projects/php-wasm/emscripten/test/report_result.h']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sean/projects/php-wasm/emscripten/test/test_browser.py", line 5585, in test_fetch_polyfill_shared_lib
    test([], expect_fail=True)
  File "/home/sean/projects/php-wasm/emscripten/test/test_browser.py", line 5577, in test
    self.compile_btest('main.c', ['-fPIC', '-Wno-unused-command-line-argument', 'library.so', '-sMAIN_MODULE=2', '-sEXIT_RUNTIME', '-o', 'a.out.html'] + args)
  File "/home/sean/projects/php-wasm/emscripten/test/common.py", line 2281, in compile_btest
    self.run_process([compiler_for(filename), filename] + self.get_emcc_args() + args)
  File "/home/sean/projects/php-wasm/emscripten/test/common.py", line 1555, in run_process
    self.fail(f'subprocess exited with non-zero return code({e.returncode}): `{shared.shlex_join(cmd)}`')
AssertionError: subprocess exited with non-zero return code(1): `/home/sean/projects/php-wasm/emscripten/emcc main.c -sNO_DEFAULT_TO_CXX -Wclosure -Werror -Wno-limited-postlink-optimizations -Wno-pointer-sign -Wno-int-conversion -fPIC -Wno-unused-command-line-argument library.so -sMAIN_MODULE=2 -sEXIT_RUNTIME -o a.out.html --pre-js /home/sean/projects/php-wasm/emscripten/test/browser_reporting.js report_result.o -include /home/sean/projects/php-wasm/emscripten/test/report_result.h`

----------------------------------------------------------------------
Ran 1 test in 0.485s

FAILED (failures=1)
```

Comment (sbc100) [reply test/common.py]: Ah thats kind of a shame.. the args include the `-o .. ` flags.   I guess you `-fPIC` approach is fine.

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Oops, I think we overlooked the webview + file-url case.   See https://github.com/emscripten-core/emscripten/pull/22026
